py/formatfloat.c

```diff
@@ -259,19 +262,46 @@ int mp_format_float(FPTYPE f, char *buf, size_t buf_size, char fmt, int prec, ch
         }
     } else {
         // Build positive exponent
-        for (e = 0, e1 = FPDECEXP; e1; e1 >>= 1, pos_pow++, neg_pow++) {
-            if (*pos_pow <= f) {
-                e += e1;
-                f *= *neg_pow;
-            }
-        }
 
-        // It can be that f was right on the edge of an entry in pos_pow needs to be reduced
-        if ((int)f >= 10) {
-            e += 1;
-            f *= FPCONST(0.1);
+        // First block is only for numbers that could be integers
+        // exactly-represented in float32.
+        if (f < FPCONST(4294967296.0)) {  // First number too big for uint32_t.
```

Comment:
Is this code all based around formatting single precision floats well?  Or does it also work for double precision?  Maybe `f_int` and `u_int` need to be `uint64_t` when `mp_float_t` is configured to be `double`?

Reply:
It works for double precision, but still only applies for numbers up to 2^32.  We could go further by extending to 64 bits, but I think the vast majority of cases of interest are covered by the existing code.

Reply:
The current code stops "working" around 2^32.  For a float32-based build:
```
>>> print("{:f}".format(4294967167))
4294967040.000000
>>> print("{:f}".format(4294967168))
4294967174.530029
```
i.e. the first number passes the test on line 268 (but then suffers precision loss since the whole value is only stored with 23 bits).  The second number, after casting into a float32, does not pass the test and is formatted according to the old code path, with the resulting fractional noise.

It would be possible to make this work for any whole-valued float but I wouldn't be able to use the trick of casting into int32s.  However, I'm disappointed by the code size results.  Maybe my original pure-float approach was better.